### PR TITLE
[InstCombine] Canonicalize `(gep <not i8> p, (div exact X, C))`

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[InstCombine] Add tests for canonicalizing `(gep <not i8> p, (exact_ins X, C))`; NFC**
- **[InstCombine] Canonicalize `(gep <not i8> p, (div exact X, C))`**

If C % sizeof(gep_element_type) is zero, we can canonicalize to `i8` via:
    `(gep i8 p, (div exact X, C / (sizeof(gep_element_type))))`

Proofs: https://alive2.llvm.org/ce/z/rHuUyh

### CODE REVIEW

**Comment by davemgreen** [comment]
Thanks for doing this! It's one thing off my todo list :)

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 1]
```diff
@@ -2939,18 +2939,57 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
...
+          std::optional<APInt> NewC;
+          if (has_single_bit(TyAllocSize) &&
+              match(ExactIns, m_Shr(m_Value(V), m_APInt(C))) &&
+              C->uge(countr_zero(TyAllocSize)))
+            NewC = *C - countr_zero(TyAllocSize);
+          else if (match(ExactIns, m_UDiv(m_Value(V), m_APInt(C)))) {
+            APInt Quot;
+            uint64_t Rem;
+            APInt::udivrem(*C, TyAllocSize, Quot, Rem);
+            if (!Quot.isAllOnes() && Rem == 0)
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2973]:
```suggestion
            if (Rem == 0)
```

**Review by dtcxzyw — APPROVED** [review]
LGTM.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp (hunk 1 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2973]: Hmm? Need the `!Quot.isAllOnes()`. See: https://alive2.llvm.org/ce/z/dSSJGA

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2973]: I don't see this check in the proof.

```
----------------------------------------
define ptr @src_udiv(ptr %p, i8 %x, i8 %C) {
#0:
  %rem = urem i8 %C, 4
  %rem_zero = icmp eq i8 %rem, 0
  assume i1 %rem_zero
  %mul = udiv exact i8 %x, %C
  %gep = gep ptr %p, 4 x i8 %mul
  ret ptr %gep
}
=>
define ptr @tgt_udiv(ptr %p, i8 %x, i8 %C) {
#0:
  %rem = urem i8 %C, 4
  %rem_zero = icmp eq i8 %rem, 0
  assume i1 %rem_zero
  %cdiv = udiv i8 %C, 4
  %mul = udiv exact i8 %x, %cdiv
  %gep = gep ptr %p, 1 x i8 %mul
  ret ptr %gep
}
Transformation seems to be correct!
```

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2973]: Bah, didn't mean to leave that in the unsigned case

File: llvm/test/Transforms/InstCombine/getelementptr.ll [hunk 2]
```diff
@@ -1790,4 +1790,106 @@ define ptr @gep_sel_const_nuw(i1 %c) {
...
+define ptr @gep_of_ashr_fail_not_divisible(ptr %p, i64 %x) {
+; CHECK-LABEL: @gep_of_ashr_fail_not_divisible(
+; CHECK-NEXT:    [[IDX:%.*]] = ashr exact i64 [[X:%.*]], 1
+; CHECK-NEXT:    [[R:%.*]] = getelementptr i32, ptr [[P:%.*]], i64 [[IDX]]
+; CHECK-NEXT:    ret ptr [[R]]
+;
+  %idx = ashr exact i64 %x, 1
+  %r = getelementptr i32, ptr %p, i64 %idx
+  ret ptr %r
+}
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/getelementptr.ll:1892]: Multi-use test?

**Review by nikic — APPROVED** [review]
LGTM

I think we should consolidate all the ptradd canonicalizations via EmitGEPOffset (i.e. only check profitability, and let everything else fold), but landing this as-is for now is fine.

**Pull request closed without merging**